mitiq/tests/test_folding_cirq.py

```diff
@@ -1,43 +1,414 @@
-import numpy as np
+"""Unit tests for folding Cirq circuits."""
+
 import random
-from cirq import Circuit, GridQubit, X, Y, Z
-from mitiq.folding_cirq import local_folding, unitary_folding
+
+import numpy as np
+import pytest
+from cirq import (Circuit, GridQubit, LineQubit, ops)
+
+from mitiq.folding_cirq import (fold_gate_at_index_in_moment,
+                                fold_gates_in_moment,
+                                fold_gates,
+                                fold_moments,
+                                fold_gates_from_left,
+                                fold_gates_at_random,
+                                fold_local,
+                                unitary_folding)
 
 
-def random_circuit(depth: int):
-    """Returns a single-qubit random circuit based on Pauli gates."""
-    # defines qubit
-    q = GridQubit(0, 0)
+STRETCH_VALS = [1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.5, 4.5, 5.0]
+DEPTH = 100
+
+
+# TODO: We will likely need to generate random circuits for testing other modules.
+#  We may want to make a (testing) utility to create random circuits.
+def random_circuit(depth: int, **kwargs):
+    """Returns a single-qubit random circuit with on Pauli gates."""
+    if "seed" in kwargs.keys():
+        np.random.seed(kwargs.get("seed"))
+    qubit = GridQubit(0, 0)
     circuit = Circuit()
     for _ in range(depth):
-        circuit += random.choice([X(q), Y(q), Z(q)])
+        circuit += random.choice([ops.X(qubit), ops.Y(qubit), ops.Z(qubit)])
     return circuit
 
 
-STRETCH_VALS = [1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.5, 4.5, 5.0]
-DEPTH = 100
+def _equal(circuit_one: Circuit, circuit_two: Circuit) -> bool:
```

Comment:
I'm sort of surprised this isn't built into Cirq. Maybe should upstream it? What do you think @rmlarose ?

Reply:
I actually think it's non-trivial to check circuit equality. I had a hunch the two checks in this function weren't completely correct, and this turns out to be true. Consider the two following circuits:

Circuit 1: H(0), H(0), H(0), X(0)
Circuit 2: H(0), X(0), X(0), X(0)

The number of operations is the same and the set of operations is the same, but the circuits aren't equal.

Fortunately, the (slightly) non-trivial way to properly check circuit equality -- by seeing if the directed acyclic graphs are isomorphic -- is already [implemented in Cirq](https://github.com/quantumlib/Cirq/blob/master/cirq/circuits/circuit_dag.py#L128). I'll update the `_equal` function to use this.

Reply:
+1 good catch